veps/sig-compute/migration-prio-queue.md

```diff
@@ -0,0 +1,121 @@
+(e.g., multiple hot plug operations) can overwhelm the system, delaying
+critical operations like node drains or KubeVirt upgrades. This proposal
+introduces a flexible priority system to ensure system operations are
+prioritized while maintaining the precedence of running migrations.
+
+## Goals
+
+- Add a `priority` field to `VirtualMachineInstanceMigrationSpec`.
+- Update the migration controller to order the queue by `priority`.
+- Assign a high, reserved priority (e.g., 1000) to running migrations.
+- Enable controllers to set priorities based on migration type.
+- Restrict user-set priorities via admission webhooks.
+- Ensure backward compatibility.
+
+## Non Goals
+
+- Change how migrations are triggered or processed.
+- Introduce new migration types.
+- Make migrations preemptible.
+
+## User Stories
+
+- As a cluster administrator, I want KubeVirt system-critical operations to proceed without delays from user actions.
+- As a VM owner, I want my hot plug operations processed efficiently, respecting system priorities.
+
+## Repos
+
+* [kubevirt/kubevirt](https://github.com/kubevirt/kubevirt)
+
+## Design
+
+A `priority` field will be added to the VMIM API, allowing various controllers to
+set priorities for migrations. The migration controller will use this field
+to order migrations in the priority queue, ensuring system-initiated migrations
+are processed before user-initiated ones. Running migrations will be get a reserved
+high priority (e.g., 1000).
+Controllers will set priorities based on the migration type, and admission webhooks will ensure users cannot
+set high priorities, protecting system operations.
+
+### Priority Order
+
+The priority order for migrations is as follows, with higher numerical values indicating higher priority:
+
+- Running Migrations: 1000 (Protected)
+- System-Critical Migrations: 100 - These include urgent tasks like node drains, evacuations, and KubeVirt upgrades. They’re critical for system health and need immediate attention.
+- User-Triggered Operations: 50 -  These are user-initiated changes, like hot plug operations.  They’re time-sensitive but should not affect system stability.
+- System Maintenance Migrations: 20 - Less urgent maintenance tasks, such as those started by the de-scheduler.
+- Manual VMIM Creation without priority or custom admin setting: This applies to admin-created VMIMs where no priority is set. They get the lowest priority by default.
+
+## API Examples
+
+```go
+type MigrationPriority int32
+
+const (
+  SystemCritical    MigrationPriority 100
+  UserTriggered     MigrationPriority 50
+  SystemMaintenance MigrationPriority 20
+)
+
+type VirtualMachineInstanceMigrationSpec struct {
+    // Priority of the migration, higher values indicate higher priority.
+    // +optional
+    Priority *MigrationPriority `json:"priority,omitempty"`
+[..]
+}
+```
+
+
+## Update/Rollback Compatibility
+
+ - VMIMs without `priority` default to 0 and the prioirity field is optional.
```

Comment:
So this basically means "vanilla" vmims get the worst priority implicitly?

I wonder if this might be problematic with current kubevirt setups where users can initiate migrations via UI for example. Is it possible that some vendors would prefer that such migrations would have higher priority than system-maintenance or even user-triggered migrations?

maybe I'm overthinking this and we can solve this with documentation, so they could simply post vmims with an explicit priority.

Reply:
Only cluster admins can initiate live migrations.
@iholder101 KubeVirt doesn't offer any UI. The administrator may choose to set a non 0 priority.